### ISSUE TITLE
Data type extensibility

### ISSUE THREAD

**Issue by spalladino** [crystal-lang/crystal-db#5]
Allow each driver to extend the data types allowed.

**Comment by asterite** [comment]
- [ ] Remove DB::Any type
- [ ] Add DB::Blob type as a struct (`record DB::Blob, bytes : Slice(UInt8)`) 
- [ ] Add `read(t : Class)` in `DB::ResultSet` to support reading any type. Base type raises
- [ ] Make sure queries (insert and query) work with any type. Base type raises on unexpected type.
- [ ] Implement this for sqlite3: https://github.com/bcardiff/crystal-sqlite3 (branch `db`)

**Comment by bcardiff** [comment]
Check read_object branch.

**Comment by bcardiff** [comment]
The DB::Any type is still useful to store params in array until `[] of Object` is supported. Although it wont be used as type restrictions, the declaration of DB::Any and DB::TYPES will remain in the source.

**Comment by bcardiff** [comment]
> Add DB::Blob type as a struct (`record DB::Blob, bytes : Slice(UInt8)`)

The motivation was to allow using just a Slice as param. That is now supported since the public api and the abstract method to be implemented differ.
Yet all Slice(UInt8) are replaced now by Bytes alias.

**Comment by bcardiff** [comment]
POC of time support at https://github.com/bcardiff/crystal-sqlite3/commit/497379ff81b509d93e0785a3e2e4c53deb837871

**Comment by bcardiff** [comment]
@asterite @spalladino I think the feature branch can be merged at db/master branch and the feature branches in sqlite/mysql to the db branch and move forward with the rest of the issues.  WDYT?

**Comment by asterite** [comment]
Yes, let's do it!

I also managed to implement the DB API for crystal-pg, it was very easy: https://github.com/asterite/crystal-pg/tree/feature/db

**Comment by spalladino** [comment]
Looks good to me!

**Comment by asterite** [comment]
I've been thinking a bit more about this and I think this could be made a lot more simpler and as (or more) powerful.

We could require `ResultSet` implementation to **just** define a `read` method that would return the next value. This will be a union of all values supported by the database. Creating a union isn't inefficient because it's just a small struct, and in any case all drivers already create a union internally, they just later cast the value.

So, the idea is to have this:

```cr
module DB
  abstract class ResultSet
    # Reads the next column value
    abstract def read
    
    # Reads the next column value as a **type**
    def read(type : Class)
      type.cast(read)
    end
  end
end
```

So, you can use it like this:

```cr
rs.read # reads a union of all types
rs.read.as(String) # reads and casts as string
rs.read.as(String?) # reads and casts as nilable string

rs.read(String) # reads as string
rs.read(String | Nil) # reads as nilable string
```

The advantage of a simple `read` method with cast is that the compiler is already pretty good at checking this, and it gives a proper error message. Otherwise every driver has to implement this logic (and in fact every driver now uses `as`, so it's kind of redundant to do that all the time).

The `read(Class)` overload is useful to read, for example, a `Time` from sqlite3. The driver could define `read(t : Time.class)`, read the value, and check its type: if it's an integer assume it's unix epoch time, etc. (this is a convention in sqlite3). So `read(Class)` is the method that should usually be used. Another example is the postgres driver, which returns a `PG::Numeric` type for numbers, but it could define `read(t : Int32.class)` to convert it to an Int32 so the user doesn't have to worry about this.

We can then define in the base ResultSet overloads for integer types like `Int32`, `Int64`, etc., which would just do `read.as(Int).to_i64`, etc. In that way all result sets don't need to care about this.

From there, I think the best thing to do it to allow creating a type from a ResultSet, similar to how you create a type from JSON:

```cr
Tuple(Int32, Int64).new(rs)
```

For primitive types we would just do `rs.read(self)`, for more complex types we can do more complex things. And with that a simple DB.mapping could be defined.

The best thing is that this is **more type safe** than the current solution, because `as` will not compile if the type is not convertible. For example doing `rs.read(Foo)` will fail to compile if no driver returns a `Foo`. In that way if you, say, use `PG::Geo::Point` in your code and later decide to migrate to mysql, you will get a compile error.

I was going to send a PR for this, but because of https://github.com/crystal-lang/crystal/issues/2916 I can't do it yet (but it's already fixed, so I'll send it after 0.18.5).

**Comment by bcardiff** [comment]
> Creating a union isn't inefficient because it's just a small struct, and in any case all drivers already create a union internally

In mysql (wihout c bindings) the `as` is needed due to how the code is organized. It access first the column type descriptor to get the actual column type and then the `as` is applied to **check** the type. But even no union is created explicitly, the call site is creating a union :-( you are right.

I would really like to be able to access the socket and return the expected value without extra operations and with the minimal memory footprint.

With the drivers implemented as they are right now, it's true what you said and the proposal sounds good for simplicity. A code refactor might allow to avoid unions all in all when calling `read(T)`, maybe with more code/metaprogramming in the driver (but I think is fine to do that).

cc: @waj
